cmd/castRunner.go

```
@@ -0,0 +1,356 @@
+			log.WithError(err).Info("Error running kubeshark tap.")
+			cancel()
+		}
+	}()
+
+	log.Info("Started kubeshark tap. ")
+
+	// Helm deploy cast
+
+	// Define a public chart repository.
+	chartRepo := repo.Entry{
+		Name: "corshatech",
+		URL:  "https://corshatech.github.io/cast/",
+	}
+
+	// Add a chart-repository to the client.
+	if err := helmClient.AddOrUpdateChartRepo(chartRepo); err != nil {
+		log.WithError(err).Info("Error adding CAST helm repo.")
+		cancel()
+	}
+
+	log.Info("Added CAST helm chart repository.")
+
+	// Install CAST chart release.
+
+	defer castCleanup(helmClient, clientset, ksPath)
+
+	go func(ctx context.Context) {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			deployCast(ctx, cancel, helmClient, clientset, config)
+		}
+	}(ctx)
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt)
+
+	// wait for signal interrupt
+	select {
+	case <-ctx.Done():
+	case <-c:
+		cancel()
+	}
+
+}
+
+// deployCast is a helper function that deploys the CAST helm chart.
+func deployCast(ctx context.Context, cancel context.CancelFunc, helmClient helm.Client, clientset *kubernetes.Clientset, config *rest.Config) {
+
+	chartSpec := helm.ChartSpec{
+		ReleaseName:     "cast",
+		CreateNamespace: true,
+		ChartName:       "corshatech/cast",
+		Namespace:       "cast",
+		UpgradeCRDs:     true,
+		Wait:            true,
+		Timeout:         2 * time.Minute, // postgres pod can take >30s to be ready.
+	}
+	if _, err := helmClient.InstallOrUpgradeChart(context.Background(), &chartSpec, nil); err != nil {
+		handleError("Error installing CAST helm chart.", err, ctx)
+		cancel()
+		return
+	}
+
+	log.Info("CAST release is ready.")
+
+	// Portforward UI pod to localhost:3000
+
+	listOptions := metav1.ListOptions{
+		LabelSelector: "app=cast-ui",
+	}
+	pods, err := clientset.CoreV1().Pods("cast").List(context.Background(), listOptions)
+	if err != nil {
+		handleError("Error listing pods in cast namespace.", err, ctx)
+		cancel()
```

Since this context is inherited it's better not to cancel it in this function. Instead we would want to return an error and let the context owner catch the error and perform cancellation as part of teardown in its defer statements. Better to return here anyway since it doesn't seem we can continue with the port forward if we failed to list pods.